irohad/multi_sig_transactions/state/impl/mst_state.cpp

```diff
@@ -122,6 +122,19 @@ namespace iroha {
     extractExpiredImpl(current_time, boost::none);
   }
 
+  void MstState::eraseByTransactionHash(
+      const shared_model::interface::types::HashType &hash) {
+    for (auto it = batches_.begin(); it != batches_.end(); ++it) {
+      auto const &transactions = it->right->transactions();
+      if (std::any_of(transactions.begin(),
+                      transactions.end(),
+                      [&hash](auto const &tx) { return hash == tx->hash(); })) {
```

Comment:
I think we should better have an index like `bimap<multiset<Hash>, multiset<BatchPtr>>` to avoid this fullscan for each committed transaction. Same for pending storage.